src/shadowbox/server/manager_service.spec.ts

```diff
@@ -110,10 +112,87 @@ describe('ShadowsocksManagerService', () => {
     });
   });
 
+  describe('changeHostname', () => {
+    const goodHostnames = [
+      "localhost","example.com","www.example.org","www.exa-mple.tw","123abc.co.uk","93.184.216.34","::0","2606:2800:220:1:248:1893:25c8:1946"
+    ];
+    for (const hostname of goodHostnames) {
+      it(`accepts ${hostname}`, (done) => {
```

Comment:
For consistency with the other tests, can we have two top-level tests: `it('accepts valid hostnames')`, `it('rejects invalid hostnames')` and iterate over the hosts inside each test?

Reply:
This way has the advantage that you get better error messages as you can tell which hostname test failed.  It helped me multiple times during development, so if anything I'd say we should change the others in a separate PR

Reply:
I find it more difficult to parse where the actual test is happening. The manager service error includes the failing hostname and that should appear in the logs.

Reply:
Done